### PR TITLE
feat: added download as csv support for table panel type

### PR DESCRIPTION
**Pull request by Vikrant2520**
### Summary

- added `download as csv` option for table panel type in dashboards

#### Related Issues / PR's

fixes https://github.com/SigNoz/engineering-pod/issues/1293

#### Screenshots

https://github.com/SigNoz/signoz/assets/54737045/13dd7c6d-6656-4e1c-a4c1-bccedbc4ed2b

### CODE REVIEW

File: frontend/src/container/GridCardLayout/WidgetHeader/index.tsx [hunk 1]
```diff
@@ -17,6 +18,8 @@ import { PANEL_TYPES } from 'constants/queryBuilder';
 import useCreateAlerts from 'hooks/queryBuilder/useCreateAlerts';
 import useComponentPermission from 'hooks/useComponentPermission';
 import history from 'lib/history';
+import { RowData } from 'lib/query/createTableColumnsFromQuery';
+import { unparse } from 'papaparse';
```
Comment (YounixM) [inline frontend/src/container/GridCardLayout/WidgetHeader/index.tsx:22]: ??

**Review by YounixM — APPROVED** [review]

**Comment by YounixM** [comment]
Build failing

File: frontend/src/container/GridCardLayout/WidgetHeader/index.tsx (hunk 1 above)
Comment (Vikrant2520) [reply frontend/src/container/GridCardLayout/WidgetHeader/index.tsx:22]: XD

File: frontend/src/container/GridCardLayout/WidgetHeader/index.tsx [hunk 2]
```diff
@@ -75,15 +80,34 @@ function WidgetHeader({
 
 	const onCreateAlertsHandler = useCreateAlerts(widget);
 
+	const onDownloadHandler = useCallback((): void => {
+		const csv = unparse(tableProcessedDataRef.current);
+		const csvBlob = new Blob([csv], { type: 'text/csv;charset=utf-8;' });
+		const csvUrl = URL.createObjectURL(csvBlob);
+		const downloadLink = document.createElement('a');
+		downloadLink.href = csvUrl;
+		downloadLink.download = `table-data.csv`;
```
Comment (srikanthccv) [inline frontend/src/container/GridCardLayout/WidgetHeader/index.tsx:89]: Let's try to use the name of the panel.

File: frontend/src/container/GridTableComponent/index.tsx [hunk 3]
```diff
@@ -27,6 +32,34 @@ function GridTableComponent({
 		[data, query],
 	);
 
+	const createDataInCorrectFormat = useCallback(
+		(dataSource: RowData[]): RowData[] =>
+			dataSource.map((d) => {
+				const finalObject = {};
+				const keys = Object.keys(d);
+				console.log(finalObject, keys);
```
Comment (srikanthccv) [inline frontend/src/container/GridTableComponent/index.tsx:40]:
```suggestion
```

**Review by srikanthccv — DISMISSED** [review]
LGTM, minor comments

File: frontend/src/container/GridCardLayout/WidgetHeader/index.tsx (hunk 2 above)
Comment (Vikrant2520) [reply frontend/src/container/GridCardLayout/WidgetHeader/index.tsx]: done

File: frontend/src/container/GridTableComponent/index.tsx (hunk 3 above)
Comment (Vikrant2520) [reply frontend/src/container/GridTableComponent/index.tsx]: removed

**Review by srikanthccv — APPROVED** [review]

File: frontend/src/container/GridCardLayout/WidgetHeader/index.tsx [hunk 4]
```diff
@@ -75,15 +81,34 @@ function WidgetHeader({
 
 	const onCreateAlertsHandler = useCreateAlerts(widget);
 
+	const onDownloadHandler = useCallback((): void => {
+		const csv = unparse(tableProcessedDataRef.current);
```
Comment (YounixM) [inline frontend/src/container/GridCardLayout/WidgetHeader/index.tsx:85]: We should add error handling here

**Review by YounixM — APPROVED** [review]

File: frontend/src/container/GridCardLayout/WidgetHeader/index.tsx (hunk 4 above)
Comment (Vikrant2520) [reply frontend/src/container/GridCardLayout/WidgetHeader/index.tsx:85]: will add everywhere this function is called in next PR

**Pull request merged**